src/components/tl/ucp/allgather/allgather_linear.c

```diff
@@ -0,0 +1,164 @@
+    task->allgather_linear.nreqs =
+        nreqs == 0 ? UCC_TL_TEAM_SIZE(tl_team) - 1 : nreqs;
+    *task_h = &task->super;
+
+    return UCC_OK;
+}
+
+/* Linear Batched K-send/receive in flight */
+ucc_status_t
+ucc_tl_ucp_allgather_linear_batched_init(ucc_base_coll_args_t *coll_args,
+                                         ucc_base_team_t      *team,
+                                         ucc_coll_task_t     **task_h)
+{
+    return ucc_tl_ucp_allgather_linear_init_batched(
+        coll_args, team, task_h,
+        get_num_reqs(ucc_derived_of(team, ucc_tl_ucp_team_t)));
+}
+
+/* Linear One-Shot version of allgather */
+ucc_status_t ucc_tl_ucp_allgather_linear_init(ucc_base_coll_args_t *coll_args,
+                                              ucc_base_team_t      *team,
+                                              ucc_coll_task_t     **task_h)
+{
+    // 0 means one-shot, team size request will be used
+    return ucc_tl_ucp_allgather_linear_init_batched(coll_args, team, task_h, 0);
+}
+
+void ucc_tl_ucp_allgather_linear_progress(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_ucp_task_t *task      = ucc_derived_of(coll_task, ucc_tl_ucp_task_t);
+    ucc_tl_ucp_team_t *team      = TASK_TEAM(task);
+    ucc_rank_t         trank     = UCC_TL_TEAM_RANK(team);
+    ucc_rank_t         tsize     = UCC_TL_TEAM_SIZE(team);
+    void              *rbuf      = TASK_ARGS(task).dst.info.buffer;
+    ucc_memory_type_t  rmem      = TASK_ARGS(task).dst.info.mem_type;
+    ucc_datatype_t     dt        = TASK_ARGS(task).dst.info.datatype;
+    size_t             count     = TASK_ARGS(task).dst.info.count;
+    size_t             data_size = (count / tsize) * ucc_dt_size(dt);
+    int                nreqs     = task->allgather_linear.nreqs;
+    int                polls     = 0;
+    void              *tmprecv, *tmpsend;
+    ucc_rank_t         peer;
+
+    while ((task->tagged.send_posted < tsize - 1 ||
+            task->tagged.recv_posted < tsize - 1) &&
+           (polls++ < task->n_polls)) {
+
+        /* Progress UCP worker */
+        ucp_worker_progress(UCC_TL_UCP_TEAM_CTX(team)->worker.ucp_worker);
+
+        /* try to send data */
+        while ((task->tagged.send_posted < tsize - 1) &&
+               ((task->tagged.send_posted - task->tagged.send_completed) <
+                nreqs)) {
+            peer    = (trank + 1 + task->tagged.send_posted) % tsize;
+            tmpsend = PTR_OFFSET(rbuf, trank * data_size);
+            /* Send my data to peer */
+            UCPCHECK_GOTO_ERR(
+                ucc_tl_ucp_send_nb(tmpsend, data_size, rmem, peer, team, task),
+                task, out);
+            polls = 0;
+        }
+
+        /* Receive peer's data at peer's offset */
+        while ((task->tagged.recv_posted < tsize - 1) &&
+               ((task->tagged.recv_posted - task->tagged.recv_completed) <
+                nreqs)) {
+            peer    = (trank + 1 + task->tagged.recv_posted) % tsize;
```

Comment:
wouldn't it be better if cycle in the reverse order for receiving? Can we have deadlocks in the current implementation?

Reply:
we should not have deadlock here, because we do send and receive. But I don't know how to proof

Reply:
fixed